File: LIB/rjtool.f90
MODULE rjtool
  IMPLICIT NONE
  CONTAINS

!---------------------------------------------------------------------
! Add ring-joining numerical characters to group strings     
!          at nodes given by rjg.                                     
!---------------------------------------------------------------------
SUBROUTINE rjgadd(nring,group,rjg)
  USE keyparameter, ONLY : digit
  IMPLICIT NONE

  INTEGER,INTENT(IN) :: nring     ! # of rings in chem 
  INTEGER,INTENT(IN) :: rjg(:,:)  ! group # of ring-joining pairs for each ring
  CHARACTER(LEN=*),INTENT(INOUT) :: group(:)  ! groups at position (node) i

  INTEGER :: n,i,ii,j

! loop in reverse order => chars in numerical order if >1 exist at any node
  DO n=nring,1,-1
    DO ii=1,2
      i = rjg(n,ii)
      IF(i/=0)THEN
        IF (group(i)(1:2)=='-O' .OR. group(i)(2:2)=='d') THEN
          j=3
        ELSE
          j=2
        ENDIF
        group(i)(j+1:)=group(i)(j:)
        group(i)(j:j) = digit(n)
      ENDIF
    ENDDO
  ENDDO
  
  END SUBROUTINE rjgadd

!-------------------------------------------------------------
! Strip ring-joining numerical characters from group strings. 
! Subroutine rjgadd puts them back.                           
!-------------------------------------------------------------
SUBROUTINE rjgrm(nring,group,rjg)
  USE keyparameter, ONLY : digit
  IMPLICIT NONE

  INTEGER,INTENT(IN)  :: nring                ! # of rings in chem
  CHARACTER(LEN=*),INTENT(INOUT) :: group(:)  ! groups at position (node) i
  INTEGER,INTENT(OUT) ::   rjg(:,:)  ! group # of ring-joining pairs for each ring

  INTEGER :: n,i,j

  rjg(:,:)=0
  DO n=1,nring
    DO i=1,SIZE(group)
      IF(group(i)(1:2)=='-O' .OR. group(i)(2:2)=='d') THEN
        j=3
      ELSE
        j=2
      ENDIF
      IF(group(i)(j:j)==digit(n)) THEN
        group(i)(j:)=group(i)(j+1:)
        IF (rjg(n,1)==0) THEN
          rjg(n,1) = i
        ELSE
          rjg(n,2) = i
        ENDIF
      ENDIF
    ENDDO
  ENDDO

END SUBROUTINE rjgrm

!-----------------------------------------------------------------
! PURPOSE: Add ring-joining numerical characters to group strings 
!          at nodes given by rj.                                  
!-----------------------------------------------------------------
SUBROUTINE rjsadd(nring,chem,rjs)
  USE keyparameter, ONLY:digit
  IMPLICIT NONE

  INTEGER, INTENT(IN) :: nring
  INTEGER, INTENT(IN) :: rjs(:,:)  ! character # of ring-joining pairs for each ring
  CHARACTER(LEN=*),INTENT(INOUT) :: chem  ! chemical formula

  INTEGER ::   n,i,j

! loop in reverse order => chars in numerical order if >1 exist at any node
  loopr: DO n=nring,1,-1
    loopp: DO j=2,1,-1
      i = rjs(n,j)
      IF (i==0) CYCLE loopp
        chem(i+1:) = chem(i:)
        chem(i:i)=digit(n)
    ENDDO loopp
  ENDDO loopr

END SUBROUTINE rjsadd

!----------------------------------------------------------------------
! PURPOSE: Strip ring-joining numerical characters from chem strings.  
!          Subroutine rjsadd puts them back.                           
!----------------------------------------------------------------------
SUBROUTINE rjsrm(nring,chem,rjs)
  USE keyparameter, ONLY: digit
  IMPLICIT NONE

  INTEGER, INTENT(IN) ::  nring          ! 
  CHARACTER(LEN=*),INTENT(INOUT) :: chem ! chemical name 
  INTEGER, INTENT(OUT) :: rjs(:,:)  ! character # of ring-joining pairs for each ring

  INTEGER  ::  n,i,j,ptr,lenchem

  rjs(:,:)=0  ; lenchem=INDEX(chem,' ')-1
  ringloop : DO n=1,nring
    pairloop: DO j=1,2
      ptr = 1
      DO i=1,lenchem
        IF(chem(i:i)=='C' .OR. chem(i:i)=='c' .OR. chem(i:i)=='-')THEN
          ptr=i+1 
          IF (chem(ptr:ptr)=='d') ptr=ptr+1
          IF (chem(ptr-1:ptr)=='-O') ptr=ptr+1
          IF (chem(ptr:ptr)==digit(n)) THEN ! found ring-joining character
            rjs(n,j)=ptr
            chem(ptr:) = chem(ptr+1:)
            CYCLE pairloop ! exit
          ENDIF
        ENDIF
      ENDDO
    ENDDO  pairloop
  ENDDO ringloop

END SUBROUTINE rjsrm

END MODULE rjtool

File: LIB/rjtool.c
/* Ring-joining tool module for GECKO-A */
#include "rjtool.h"
#include "keyparameter.h"
#include <string.h>
#include <stdio.h>

/* Add ring-joining numerical characters to group strings */
void rjgadd(int nring, char **group, int **rjg, int group_size, int rjg_size) {
    int n, i, ii, j;
    char temp[256];
    
    /* loop in reverse order => chars in numerical order if >1 exist at any node */
    for (n = nring - 1; n >= 0; n--) {
        for (ii = 0; ii < 2; ii++) {
            i = rjg[n][ii] - 1;  /* Convert to 0-indexed */
            if (i >= 0) {
                if ((strlen(group[i]) >= 2 && strncmp(group[i], "-O", 2) == 0) ||
                    (strlen(group[i]) >= 2 && group[i][1] == 'd')) {
                    j = 2;  /* Position 3 in Fortran = index 2 in C */
                } else {
                    j = 1;  /* Position 2 in Fortran = index 1 in C */
                }
                
                /* Insert digit at position j */
                strcpy(temp, group[i]);
                temp[j] = gecko_digit[n];
                strcpy(temp + j + 1, group[i] + j);
                strcpy(group[i], temp);
            }
        }
    }
}

/* Strip ring-joining numerical characters from group strings */
void rjgrm(int nring, char **group, int **rjg, int group_size, int rjg_size) {
    int n, i, j;
    
    /* Initialize rjg to 0 */
    for (n = 0; n < rjg_size; n++) {
        rjg[n][0] = 0;
        rjg[n][1] = 0;
    }
    
    for (n = 0; n < nring; n++) {
        for (i = 0; i < group_size; i++) {
            if ((strlen(group[i]) >= 2 && strncmp(group[i], "-O", 2) == 0) ||
                (strlen(group[i]) >= 2 && group[i][1] == 'd')) {
                j = 2;  /* Position 3 in Fortran = index 2 in C */
            } else {
                j = 1;  /* Position 2 in Fortran = index 1 in C */
            }
            
            if (group[i][j] == gecko_digit[n]) {
                /* Remove the digit by shifting the string */
                memmove(group[i] + j, group[i] + j + 1, strlen(group[i] + j + 1) + 1);
                
                if (rjg[n][0] == 0) {
                    rjg[n][0] = i + 1;  /* Store as 1-indexed */
                } else {
                    rjg[n][1] = i + 1;  /* Store as 1-indexed */
                }
            }
        }
    }
}

/* Add ring-joining numerical characters to chem string */
void rjsadd(int nring, char *chem, int **rjs, int rjs_size) {
    int n, i, j;
    char temp[256];
    
    /* loop in reverse order => chars in numerical order if >1 exist at any node */
    for (n = nring - 1; n >= 0; n--) {
        for (j = 1; j >= 0; j--) {
            i = rjs[n][j] - 1;  /* Convert to 0-indexed */
            if (i >= 0 && i < (int)strlen(chem)) {
                /* Insert digit at position i */
                strcpy(temp, chem);
                temp[i] = gecko_digit[n];
                strcpy(temp + i + 1, chem + i);
                strcpy(chem, temp);
            }
        }
    }
}

/* Strip ring-joining numerical characters from chem string */
void rjsrm(int nring, char *chem, int **rjs, int rjs_size) {
    int n, i, j, ptr, lenchem;
    
    /* Initialize rjs to 0 */
    for (n = 0; n < rjs_size; n++) {
        rjs[n][0] = 0;
        rjs[n][1] = 0;
    }
    
    /* Find length of chem (up to first space or end) */
    lenchem = strcspn(chem, " ");
    if (lenchem == 0) lenchem = strlen(chem);
    
    for (n = 0; n < nring; n++) {
        for (j = 0; j < 2; j++) {
            ptr = 0;
            for (i = 0; i < lenchem; i++) {
                if (chem[i] == 'C' || chem[i] == 'c' || chem[i] == '-') {
                    ptr = i + 1;
                    if (ptr < lenchem && chem[ptr] == 'd') ptr++;
                    if (ptr >= 2 && strncmp(chem + ptr - 2, "-O", 2) == 0) ptr++;
                    
                    if (ptr < lenchem && chem[ptr] == gecko_digit[n]) {
                        rjs[n][j] = ptr + 1;  /* Store as 1-indexed */
                        /* Remove the digit */
                        memmove(chem + ptr, chem + ptr + 1, strlen(chem + ptr + 1) + 1);
                        lenchem--;
                        break;
                    }
                }
            }
        }
    }
}
